lib/fizzy/instantiate.hpp

```diff
@@ -21,23 +21,79 @@ namespace fizzy
 struct ExecutionResult;
 struct Instance;
 
-/// Function representing WebAssembly or host function execution.
-using execute_function = std::function<ExecutionResult(Instance&, const Value*, int depth)>;
+/// Function pointer to the execution function.
+using ExecuteFunctionPtr = ExecutionResult (*)(
+    void* context, Instance&, const Value*, int depth) noexcept;
+
+/// Function representing either WebAssembly or host function execution.
+class ExecuteFunction
+{
+    /// Pointer to WebAssembly function instance.
+    /// Equals nullptr in case this ExecuteFunction represents host function.
+    Instance* m_instance = nullptr;
+
+    /// Index of WebAssembly function.
+    /// Equal 0 in case this ExecuteFunction represents host function.
+    FuncIdx m_func_idx = 0;
+
+    /// Pointer to a host function.
+    /// Equals nullptr in case this ExecuteFunction represents WebAssembly function.
+    ExecuteFunctionPtr m_host_function = nullptr;
+
+    /// Opaque context of host function execution, which is passed to it as a context parameter.
+    /// `shared_ptr` is used because ExecuteFunction can be copied.
+    /// Empty in case this ExecuteFunction represents WebAssembly function.
+    std::shared_ptr<void> m_host_context;
+
+public:
+    /// Default function constructor.
+    ExecuteFunction() noexcept = default;
+    /// Copy constructor.
+    ExecuteFunction(const ExecuteFunction&) noexcept = default;
```

Comment:
None of these 4 declarations are strictly needed. They will be added anyway if you don't declare a single one of them nor destructor.

Reply:
This is correct, I wanted to make them explicit, because they are used (not all of them though.)

Reply:
See [Rule of zero](https://en.cppreference.com/w/cpp/language/rule_of_three#Rule_of_zero).